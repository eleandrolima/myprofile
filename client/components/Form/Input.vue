<template>
  <div>
    <label class="label">
      <span class="label-text">{{ label }}</span>
    </label>
    <input
      :placeholder="placeholder"
      class="input input-bordered w-full"
      :class="computedClass"
      :type="type"
      @input="$emit('input', $event.target.value)"
    >
    <div class="tracking-wide text-red-500 text-xs mt-1 ml-1">
      <div v-for="violation in violations">
        {{ violation.message }}
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { ViolationInterface } from '~/exception/contracts/violation.interface'

export default {
  name: 'Input',
  props: {
    type: { type: String, default: 'text' },
    placeholder: { type: String, required: true },
    label: { type: String, required: true },
    violations: { type: Array as () => ViolationInterface[], required: true }
  },
  computed: {
    computedClass () {
      if (this.violations.length) {
        return 'input-error'
      }

      return ''
    }
  }
}
</script>
