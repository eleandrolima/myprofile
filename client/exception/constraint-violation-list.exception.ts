import { ViolationInterface, ViolationsInterface } from '~/exception/contracts/violation.interface'

export class ConstraintViolationListException implements ViolationsInterface {
  violations: ViolationInterface[] = []

  constructor (init?: Partial<ViolationsInterface>) {
    Object.assign(this, init)
  }

  public getViolationsBy (property: string): ViolationInterface[] {
    return this.violations.filter((violation: ViolationInterface) => violation.propertyPath == property)
  }
}
