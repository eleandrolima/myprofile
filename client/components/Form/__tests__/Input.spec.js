import { shallowMount } from '@vue/test-utils'
import Input from '../Input'
import { ConstraintViolationListException } from '~/exception/constraint-violation-list.exception'
import json from '~/static/__mocks__/violations'

describe('Input component', () => {
  test('props are working properly.', () => {
    const wrapper = shallowMount(Input, {
      propsData: {
        placeholder: 'foo',
        type: 'bar',
        label: 'baz',
        violations: []
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('default props.', () => {
    const wrapper = shallowMount(Input, {
      propsData: {
        placeholder: 'foo',
        label: 'baz',
        violations: []
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can see error message.', () => {
    const wrapper = shallowMount(Input, {
      propsData: {
        placeholder: 'foo',
        label: 'baz',
        violations: new ConstraintViolationListException(JSON.parse(json)).violations
      }
    })

    expect(wrapper.find('#violations').text()).toBe('baz message.')
  })
})
