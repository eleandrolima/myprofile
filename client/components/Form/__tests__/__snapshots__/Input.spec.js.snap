// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component default props. 1`] = `<div><label class="label"><span class="label-text">baz</span></label> <input placeholder="foo" type="text" class="input input-bordered w-full"></div>`;

exports[`Input component props are working properly. 1`] = `<div><label class="label"><span class="label-text">baz</span></label> <input placeholder="foo" type="bar" class="input input-bordered w-full"></div>`;
